### ISSUE TITLE
Discrepancy between API and documentation with `.observer`

### ISSUE THREAD

**Issue by galou** [mnesarco/fcapi#2]
The `fpo` argument is missing in all function with the `@{prop}.observer` decorator.

Here is a patch to update the documentation but I'd rather update the API to remove `fpo` as argument since it can be  accessed via `self.Object` (by the way, the documentation doesn't state that).

```diff
diff --git a/docs/documentation.md b/docs/documentation.md
index 231bed0..5c205e7 100644
--- a/docs/documentation.md
+++ b/docs/documentation.md
@@ -652,7 +652,7 @@ class CustomBoxProxy:
         fpo.Shape = Part.makeBox(self.length, self.width, self.height)
     
     @length.observer
-    def length_changed(self, new_value, old_value):
+    def length_changed(self, fpo, new_value, old_value):
         print(f"Hey! length has changed from {old_Value} to {new_value}")
 
 ```
@@ -719,7 +719,7 @@ class MyMagicProxy:
 
     # Optional listener
     @my_property.observer
-    def my_property_obs(self, new_value, old_value):
+    def my_property_obs(self, fpo, new_value, old_value):
         print(f"my_property has changed from {old_value} to {new_value}")
         
 ```
@@ -809,15 +809,15 @@ class MyMagicProxy
     my_prop3 = PropertyInteger(section="Basic", default=5)
 
     @my_prop1.observer
-    def listener1(self, new_value, old_value):
+    def listener1(self, fpo, new_value, old_value):
         print(f"my_property1 has changed from {old_value} to {new_value}")
         
     @my_prop2.observer
-    def listener2(self, new_value):
+    def listener2(self, fpo, new_value):
         print(f"my_property2 has changed {new_value}")
 
     @my_prop3.observer
-    def listener3(self):
+    def listener3(self, fpo):
         print(f"my_property3 has changed")
         
 ```
@@ -1640,4 +1640,4 @@ ex10.create_cube_pd()
 ```
 
 > [!IMPORTANT]
-> Copy `fpo.py` in __only one place__ to avoid a name conflicts.
\ No newline at end of file
+> Copy `fpo.py` in __only one place__ to avoid a name conflicts.
```

**Comment by mnesarco** [comment]
@galou  Thank you.

**Comment by mnesarco** [comment]
I am not sure about removing it. It will break some of my existing code that depends on fpo. It is apparently redundant and it is inconsistent with the observers in the ViewProxy API. Let me evaluate the change for a few days before taking action.

**Comment by galou** [comment]
If removal, then I'd would remove it from all `on_*` callbacks (also in `view_proxy` naturally). I actually never understood why the `object` argument exists in callbacks in the FreeCAD API.

I understand that a lot of changes may be implied but a good regex or some AI should solve this (I can even assist in this). From my point of view this is worth it.

Are you aware of workbenches using this great API. I just started and love it. It makes writing workbenches much more pleasant. Before that, I also used a [base class](https://github.com/galou/freecad.cross/blob/85baf10e6219055c369dd48d0b72c46a8b5845bc/freecad/cross/freecad_utils.py#L603)  for proxies but the functionalities were limited to ensuring that all required properties are defined in `execute()`.

**Comment by mnesarco** [comment]
This API is not official and to be honest, there is not much interest on it from the core FreeCAD teams right now as they are very busy with FC 1.0 Release. So I don't think any existing workbench is using it right now. I started creating this API for my own internal Workbenches but finally decided to document it and publish, during the process i redesigned it almost entirely :D. My main goal is to make it consistent and documented.

I am glad you are interested, i have no problem introducing changes to make it better. I can keep my internal projects isolated from this public version. 

I paused the work on this until FC 1.0 release because current FC dev version is very unstable. The base FC API is already very inconsistent, incomplete and mostly undocumented or the documentation is very outdated, so i try to get the real info from the FC source code itself and document by myself.

I also have another API to simplify the GUI code. It was initially released as fcscript but at the end i am going to merge both projects. Also the fcscrip ui part was highly updated during my overhaull of Marz Workbech but the updated version is still used only in Marz workbench itself.

**Comment by mnesarco** [comment]
> What about removing the `fpo` argument? I can try a PR if you wish.

I have been thinking about this, and I think that is is not a good idea to relay on `self.Object`. As a general rule of thumbs it is better to avoid direct state access from extension APIs. I know that it is not 100% consistent right now, but it is always better to provide access through methods or explicit arguments rather than direct object attributes, specially when the extension is based on meta-programming and even in inheritance based extensions.

Direct attribute access to framework provided state is always risky and limits the flexibility of new unexpected changes, while methods can maintain backwards compatibility for longer periods and explicit method arguments in callbacks are more simple in many cases (not always but mostly).

As the FreeCAD python API is not exactly a well designed and consistent API, i want to keep the maximum flexibility possible to support unexpected limitations in the future.

Passing the FPO instance to the callback/listener makes it easy to delegate to external components without access to the proxy class and makes future changes easier, specially if more meta-programming is required at some point.

**Comment by galou** [comment]
I don't understand your point. Inside callbacks, I mostly rely on the bindings created by the proxy decorator (for example `self.position` for the property `fp.Position`) and `self.Object` is also a binding generated by the proxy decorator.

I'm not so confident in meta-programming, why is it a problem to access to the proxy class in a callback defined in this proxy class?

**Comment by mnesarco** [comment]
@galou There is no problem using **self.Object**, this project is in very early stages, I have get no feedback apart from you, so many aspects of it can change over time, I try to keep it not too divergent from raw API callback signatures.

The core FreeCAD team is very skeptical about my projects, they provide zero feedback, basically ignore them entirely and their process does not care too much about backwards compatibility, so there is a high risk that something change unexpectedly and undocumented in particular in lifecycle related things.

I am open to ideas if we discuss they and ponder pros and cons. At this point all decisions are arbitrary, but they will weight over time.

From August (date of your comment) to today (Dec 3th) i am more confident with relying on **self.Object**, so it appears to be a good decision from the beginning. Now your concern about signatures passing the DocumentObject as an argument, what is the worst thing about it?

**Comment by mnesarco** [comment]
Yes it is redundant for sure. I tend to think about the proxies as hollow things (proxy pattern) but with the addition of the Property bindings they looks like stateful objects. Removing fp from signatures is an easy change but I have no idea how many developers are using fpo right now and it is a breaking change. Let me think about how to manage the change process, this is very new so there is no actual semantic versioning or deprecation process. I need to define those.

**Comment by galou** [comment]
I'd happy to help migrate also your workbenches because I think that the change is worth it.

You won't break anything in other people's workbenches since they probably have their own copy of `fpo.py` and don't need to upgrade if they don't want to.

**Comment by mnesarco** [comment]
@galou 
I will introduce a different approach here to make it more flexible so i don't have to change the signatures again in the future, I want to do this change once and only once. So I will introduce Event classes everywhere and all signatures will receive a specific Event and no more arguments. this way Event classes can be enhanced over time without breaking existing code.

Example (pseudocode):

```python

class PropertyChangeEvent[T]:
  old_value: T
  new_value: T
  target: DocumentObject


@my_prop.observer
def listener3(self, event: PropertyChangeEvent):
  print(f"my_property3 has changed to {event.new_value}")


```

I will introduce the same approach for everything `on_execute`, `on_init`, ....

This also makes signatures easier to remember.

**Comment by mnesarco** [comment]
Also the event will be optional in the signature

**Comment by mnesarco** [comment]
Moving discussion to its own Issue: https://github.com/mnesarco/fcapi/issues/15
